currencies/rate_converter.go

```diff
@@ -0,0 +1,205 @@
+package currencies
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync/atomic"
+	"time"
+)
+
+// RateConverter holds the currencies conversion rates dictionnary
+type RateConverter struct {
+	httpClient            httpClient
+	updates               chan *Rates
+	done                  chan bool
+	updateNotifier        chan<- int
+	ticker                *time.Ticker
+	fetchingInterval      time.Duration
+	retryFetchingInterval time.Duration
+	syncSourceURL         string
+	rates                 atomic.Value // Should only hold Rates struct
+	lastUpdated           atomic.Value // Should only hold time.Time
+}
+
+// NewRateConverter returns a new RateConverter
+func NewRateConverter(
+	httpClient httpClient,
+	syncSourceURL string,
+	fetchingInterval time.Duration,
+	retryFetchingInterval time.Duration,
+) *RateConverter {
+	return NewRateConverterWithNotifier(
+		httpClient,
+		syncSourceURL,
+		fetchingInterval,
+		retryFetchingInterval,
+		nil, // no notifier channel specified, won't send any notifications
+	)
+}
+
+// NewRateConverterWithNotifier returns a new RateConverter
+// it allow to pass an update chan in which the number of ticks will be passed after each tick
+// allowing clients to listen on updates
+// Do not use this method
+func NewRateConverterWithNotifier(
+	httpClient httpClient,
+	syncSourceURL string,
+	fetchingInterval time.Duration,
+	retryFetchingInterval time.Duration,
+	updateNotifier chan<- int,
+) *RateConverter {
+	rc := &RateConverter{
+		httpClient:            httpClient,
+		updates:               make(chan *Rates),
+		done:                  make(chan bool),
+		updateNotifier:        updateNotifier,
+		fetchingInterval:      fetchingInterval,
+		retryFetchingInterval: retryFetchingInterval,
+		ticker:                nil,
+		syncSourceURL:         syncSourceURL,
+		rates:                 atomic.Value{},
+		lastUpdated:           atomic.Value{},
+	}
+
+	// In case host do not want to support currency lookup
+	// we just stop here and do nothing
+	if rc.fetchingInterval == time.Duration(0) {
+		return rc
+	}
+
+	rc.Update()                // Make sure to populate data before to return the converter
+	rc.startPeriodicFetching() // Start periodic ticking
+
+	return rc
+}
+
+// fetch allows to retrieve the currencies rates from the syncSourceURL provided
+func (rc *RateConverter) fetch() (*Rates, error) {
+	request, err := http.NewRequest("GET", rc.syncSourceURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	request = request.WithContext(ctx)
+
+	response, err := rc.httpClient.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	bytesJSON, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedRates := &Rates{}
+	err = json.Unmarshal(bytesJSON, &updatedRates)
```

Comment:
`updatedRates` is already a pointer, you don't want a pointer to a pointer

Reply:
Oh yes, fixed !